### PR TITLE
fix latex

### PR DESCRIPTION
**Pull request by bclavie**
Congratulations! You've made it this far! Once merged, the article will appear at https://huggingface.co/blog. Official articles 
require additional reviews. Alternatively, you can write a community article following the process [here](https://huggingface.co/blog-explorers).

## Preparing the Article

You're not quite done yet, though. Please make sure to follow this process (as documented [here](https://github.com/huggingface/blog/tree/main?tab=readme-ov-file#how-to-write-an-article-)):

- [x] Add an entry to [_blog.yml](https://github.com/huggingface/blog/blob/main/_blog.yml).
- [x] Add a thumbnail. There are no requirements here, but there is a [template](https://github.com/huggingface/blog?tab=readme-ov-file#how-to-get-a-nice-responsive-thumbnail) if it's helpful.
- [x] Check you use a short title and blog path.
- [x] Upload any additional assets (such as images) to the Documentation Images [repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog). This is to reduce bloat in the GitHub base repo when cloning and pulling. Try to have small images to avoid a slow or expensive user experience.
- [x] Add metadata (such as authors) to your `md` file. You can also specify `guest` or `org` for the authors.
- [x] Ensure the publication date is correct.
- [x] Preview the content. A quick way is to paste the markdown content in https://huggingface.co/new-blog. **Do not click publish**, this is just a way to do an early check.

Here is an example of a complete PR: https://github.com/huggingface/blog/pull/2382

## Getting a Review

Please make sure to get a review from someone on your team or a co-author.
Once this is done and once all the steps above are completed, you should be able to merge.
There is no need for additional reviews if you and your co-authors are happy and meet all of the above.

Feel free to add @pcuenca as a reviewer if you want a final check. Keep in mind he'll be biased toward light reviews
(e.g., check for proper metadata) rather than content reviews unless explicitly asked.

### CODE REVIEW

**Comment by bclavie** [comment]
@ariG23498 @pcuenca

File: modernbert.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+While padding solves the problem, it doesn’t do so elegantly: a lot of compute ends up being spent and wasted on padding tokens, which do not contribute any semantic information.
+
+![][modernbert_unpadding]  
+*Comparing padding with sequence packing. Sequence packing (‘unpadding’) avoids wasting compute on padding tokens and has more consistent non-padding token counts per batch. Samples are still processed individually through careful masking.*
+
+**Unpadding** solves this issue: rather than keeping these padding tokens, we remove them all, and concatenate them into mini-batches with a batch size of one, avoiding all unnecessary computations. If you’re using Flash Attention, our implementation of unpadding is even faster than previous methods, which heavily relied on unpadding and repadding sequences as they went through the model: we go one step further by introducing our own implementation of unpadding, relying heavily on recent developments in Flash Attention’s RoPE support. This allows ModernBERT to only have to unpad once, and optionally repad sequences after processing, resulting in a 10-20% speedup over previous methods.
```
Comment (pcuenca) [inline modernbert.md:194]: !!

File: modernbert.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+The impact of this is immediately noticeable: out of all the existing open source encoders, ModernBERT is in a class of its own on programming-related tasks. We’re particularly interested in what downstream uses this will lead to, in terms of improving programming assistants.
+
+#### Process
+
+We stick to the original BERT’s training recipe, with some slight upgrades inspired by subsequent work: we remove the Next-Sentence Prediction objective, since then shown to add overhead for no clear gains, and increase the masking rate from 15% to 30%.
+
+Both models are trained with a **three-phase process**. First, we train on 1.7T tokens at a sequence length of 1024\. We then adopt a long-context adaptation phase, training on 250B tokens at a sequence length of 8192, while keeping the total tokens seen per batch more or less consistent by lowering the batch size. Finally, we perform annealing on 50 billion tokens sampled differently, following the long-context extension ideal mix highlighted by [ProLong](https://arxiv.org/abs/2410.02660).
```
Comment (pcuenca) [inline modernbert.md:235]:
```suggestion
Both models are trained with a **three-phase process**. First, we train on 1.7T tokens at a sequence length of 1024. We then adopt a long-context adaptation phase, training on 250B tokens at a sequence length of 8192, while keeping the total tokens seen per batch more or less consistent by lowering the batch size. Finally, we perform annealing on 50 billion tokens sampled differently, following the long-context extension ideal mix highlighted by [ProLong](https://arxiv.org/abs/2410.02660).
```

**Review by pcuenca — COMMENTED** [review]
It's an amazing read! 🔥 I just suggested minor nits that can be safely ignored.

File: modernbert.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  org: answerdotai
+- user: johnowhitaker
+  guest: true
+  org: answerdotai
+- user: iacolippo
+  guest: true
+  org: lightonai
+---
+
+# Finally, a Replacement for BERT
```
[the remark this answers is not in the document]
Comment (bclavie) [reply modernbert.md:46]: I'm happy either way, not sure which one looks better on the most screens! I added the longer one to the metadata because I noticed most BPs seem to have the model name in there no matter the title.

File: modernbert.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+Of course, the open-ended capabilities of these giant generative models mean that you can, in a pinch, press them into service for non-generative or *discriminative* tasks, such as classification. This is because you can describe a classification task in plain English and ... just ask the model to classify. But while this workflow is great for prototyping, you don’t want to pay prototype prices once you’re in mass production.
+
+The popular buzz around GenAI has obscured the role of *encoder-only models*. These are the workhorses of practical language processing, the models that are actually being used for such workloads right now in many scientific and commercial applications.
```
[the remark this answers is not in the document]
Comment (bclavie) [reply modernbert.md:78]: I'm tapping out for today, but I'll ask around if someone can add a couple if you think this is really necessary! I was worried having examples this early would be redundant with the next section, where we go into more depth about a very specific one (the FineWeb-Edu one)

**Comment by bclavie** [comment]
Thanks for the kind words 😄 And thank you for spotting all the escaping! I removed a ton of it that the gdocs -> md conversion added, but it seems like I've missed a good few.

File: modernbert.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply modernbert.md:78]: Absolutely not necessary, I was thinking more around the lines of practical uses in other industries. Let's keep it like it is then, and great point about FineWeb-Edu!

File: modernbert.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+Finally, 6 years later, we have a replacement! Today, we at [Answer.AI](http://Answer.AI) and [LightOn](https://www.lighton.ai/) (and friends!) are releasing ModernBERT. ModernBERT is a new model series that is a Pareto improvement over BERT and its younger siblings across both **speed** and **accuracy**. This model takes dozens of advances from recent years of work on large language models (LLMs), and applies them to a BERT-style model, including updates to the architecture and the training process.
+
+![][modernbert_pareto_curve]
```
Comment (pcuenca) [inline modernbert.md:60]:
```suggestion
![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/modernbert/modernbert_pareto_curve.png)
```

File: modernbert.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+## Performance
+
+### Overview
+
+Here’s a snapshot of the accuracy of ModernBERT and other models across a range of tasks, as measured by standard academic benchmarks – as you can see, ModernBERT is the only model which is a **top scorer across every category**, which makes it the one model you can use for all your encoder-based tasks:  
+
+![][modernbert_accuracy_table]
```
Comment (pcuenca) [inline modernbert.md:110]:
```suggestion
![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/modernbert/modernbert_accuracy_table.png)  
```

File: modernbert.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+Compared to the mainstream models, ModernBERT performs better across nearly all three broad task categories of retrieval, natural language understanding, and code retrieval. Whilst it slightly lags [DeBERTaV3](https://huggingface.co/microsoft/deberta-v3-base) in one area (natural language understanding), it is many times faster.
+
+Compared to the specialized models, ModernBERT is comparable or superior in most tasks. In addition, ModernBERT is faster than most models across most tasks, and can handle inputs up to 8,192 tokens, 16x longer than the mainstream models. 
+
+### Efficiency
+
+Here’s the memory (max batch size, BS) and Inference (in thousands of tokens per second) efficiency results on an NVIDIA RTX 4090 for ModernBERT and other decoder models:  
+
+![][modernbert_efficiency_table]
```
Comment (pcuenca) [inline modernbert.md:130]:
```suggestion
![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/modernbert/modernbert_efficiency_table.png)
```

File: modernbert.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+As with most things in research, we build while standing on the shoulders of giants, and heavily leverage Flash Attention 2’s speed improvements. Our efficiency improvements rely on three key components: **Alternating Attention**, to improve processing efficiency, **Unpadding and Sequence Packing**, to reduce computational waste, and **Hardware-Aware Model Design**, to maximise hardware utilization.
+
+#### Global and Local Attention
+
+One of ModernBERT’s most impactful features is **Alternating** **Attention**, rather than full global attention. In technical terms, this means that our attention mechanism only attends to the full input every 3 layers (**global attention**), while all other layers use a sliding window where every token only attends to the 128 tokens nearest to itself (**local attention)**.   
+As attention’s computational complexity balloons up with every additional token, this means ModernBERT can process long input sequences considerably faster than any other model.
+
+In practice, it looks like this:  
+![][modernbert_alternating_attention]
```
Comment (pcuenca) [inline modernbert.md:179]:
```suggestion
![](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/modernbert/modernbert_alternating_attention.png)
```

File: modernbert.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+#### Unpadding and Sequence Packing
+
+Another core mechanism contributing to ModernBERT’s efficiency is its use for Unpadding and Sequence packing.
+
+In order to be able to process multiple sequences within the same batch, encoder models require them to be the *same length*, so they can perform parallel computation. Traditionally, we’ve relied on **padding** to achieve this: figure out which sentence is the longest, and add meaningless tokens (*padding tokens*) to fill up every other sequence.
+
+While padding solves the problem, it doesn’t do so elegantly: a lot of compute ends up being spent and wasted on padding tokens, which do not contribute any semantic information.
+
+![][modernbert_unpadding]  
+*Comparing padding with sequence packing. Sequence packing (‘unpadding’) avoids wasting compute on padding tokens and has more consistent non-padding token counts per batch. Samples are still processed individually through careful masking.*
```
Comment (pcuenca) [inline modernbert.md:192]:
```suggestion
<figure class="image text-center">
  <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/modernbert/modernbert_unpadding.png" alt="Padding vs sequence packing">
  <figcaption>Comparing padding with sequence packing. Sequence packing (‘unpadding’) avoids wasting compute on padding tokens and has more consistent non-padding token counts per batch. Samples are still processed individually through careful masking.</figcaption>
</figure>
```

Adding a caption, otherwise I think the paragraph would be rendered disconnected from the figure.

**Review by pcuenca — COMMENTED** [review]
Adding images

**Review by pcuenca — APPROVED** [review]
Approving to unblock, most everything is ready

File: _blog.yml [hunk 10]
```diff
@@ -5202,4 +5202,15 @@
     - nlp
     - llm
     - community
+    - research
+
```
Comment (pcuenca) [inline _blog.yml:5206]:
```suggestion

- local: bamba
  title: "Bamba: Inference-Efficient Hybrid Mamba2 Model"
  author: rganti
  guest: true
  thumbnail: /blog/assets/bamba/bamba_thumbnail.png
  date: Dec 18, 2024
  tags:
    - research
    - nlp
    - community
    - model

```

Attempt to fix conflict.

File: modernbert.md [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+
+ModernBERT demonstrates that encoder-only models can be improved by modern methods. They continue to offer very strong performance on some tasks, providing an extremely attractive size/performance ratio.
+
+More than anything, we’re really looking forward to seeing what creative ways to use these models the community will come up with! To encourage this, we’re opening a call for demos until January 10th, 2025: the 5 best ones will get added to this post in a showcase section and win a $100 (or local currency equivalent) Amazon gift card, as well as a 6-month HuggingFace Pro subscription! (If you need a hint to get started, here’s a demo we thought about: code similarity HF space)
+
+## Links
+
+- [🤗ModernBERT-Base](https://huggingface.co/answerdotai/ModernBERT-base)  
+- [🤗ModernBERT-Large](https://huggingface.co/answerdotai/ModernBERT-large)  
+- [📝**arXiv**](https://arxiv.org/abs/2412.09588)
```
Comment (pcuenca) [inline modernbert.md:265]:
```suggestion
- [📝**arXiv**](https://arxiv.org/abs/2412.13663)  
```

Comment (bclavie) [reply modernbert.md:265]: Oh wow this is one of the worst copy pastes in the history of copy pastes, thanks for catching it!

File: modernbert.md [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+#### def data(): return \[‘text’, ‘bad\_text’, ‘math’, ‘code’\]
+
+[https://media1.tenor.com/m/xJSM2Ky3WpgAAAAd/steve-ballmer-microsoft.gif](https://media1.tenor.com/m/xJSM2Ky3WpgAAAAd/steve-ballmer-microsoft.gif)  
+*Picture this exact scene, but replace Developers with Data*
+
+Another big aspect in which encoders have been trailing behind is training data. This is often understood to mean solely training data **scale**, but this is not actually the case: previous encoders, such as DeBERTaV3, were trained for long enough that they might have even breached the trillion tokens scale\!
+
+The issue, rather, has been training data **diversity**: many of the older models train on limited corpora, generally consisting of Wikipedia and Wikibooks. These data mixtures are very noticeably **single text modality**: they contain nothing but high-quality natural text.
+
+In contrast, ModernBERT is trained on data from a variety of English data, including web documents, code, and scientific articles. It is trained on **2 trillion tokens**, of which most are unique, rather than the standard 20-to-40 repetitions common in previous encoders.
```
Comment (tomaarsen) [inline modernbert.md:229]:
```suggestion
In contrast, ModernBERT is trained on data from a variety of English sources, including web documents, code, and scientific articles. It is trained on **2 trillion tokens**, of which most are unique, rather than the standard 20-to-40 repetitions common in previous encoders.
```
cc @jph00

File: modernbert.md [hunk 13]
```diff
@@ -0,0 +1,268 @@
+---
+title: "Finally, a Replacement for BERT: Introducing ModernBERT" 
+thumbnail: /blog/assets/modernbert/thumbnail.png
+authors:
+- user: bwarner 
+  guest: true
+  org: answerdotai
+- user: Nohtow
```
Comment (NohTow) [inline modernbert.md:8]:
```suggestion
- user: NohTow
```

File: modernbert.md [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+- The final embedding sizes must match the original BERT’s dimensions, 768 for base and 1024 for large, to maximize backwards compatibility  
+- Set performance constraints which are common across the basket of GPUs
+
+Afterwards, we experimented with multiple model designs via a constrained grid search, varying both layer counts and layer width. Once we’d identified shapes that appeared to be the most efficient ones, we confirmed that our heuristics matched real-world GPU performance, and settled on the final model designs.
+
+### Training
+
+#### def data(): return \[‘text’, ‘bad\_text’, ‘math’, ‘code’\]
+
+[https://media1.tenor.com/m/xJSM2Ky3WpgAAAAd/steve-ballmer-microsoft.gif](https://media1.tenor.com/m/xJSM2Ky3WpgAAAAd/steve-ballmer-microsoft.gif)
```
[the remark this answers is not in the document]
Comment (bclavie) [reply modernbert.md]: I don't know about **We**, but **I** definitely want it to embed 😂

File: modernbert.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+
+While decoder-only models (like a GPT) can do the work of an encoder-only model (like a BERT), they are hamstrung by a key constraint: since they are *generative models*, they are mathematically “not allowed” to “peek” at later tokens. They can only ever *look backwards*. This is in contrast to encoder-only models, which are **trained so each token can look forwards *and* backwards (bi-directional)**. They are built for this, and it makes them very efficient at what they do.
+
+Basically, a frontier model like OpenAI's O1 is like a Ferrari SF-23. It’s an obvious triumph of engineering, designed to win races, and that’s why we talk about it. But it takes a special pit crew just to change the tires and you can’t buy one for yourself. In contrast, a BERT model is like a Honda Civic. It’s *also* an engineering triumph, but more subtly, since *it* is engineered to be affordable, fuel-efficient, reliable, and extremely useful. And that’s why they’re absolutely everywhere.
```
[the remark this answers is not in the document]
Comment (davanstrien) [reply modernbert.md:86]: love this

File: modernbert.md [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+
+And indeed, it turns out that tiling ModernBERT-base’s weights across ModernBERT-large works better than initializing from random weights. It also has the added benefit of stacking nicely with batch size warmup for even faster initial training.
+
+## Conclusion
+
+In this blog post we introduced the ModernBERT models, a new state-of-the-art family of small and efficient encoder-only models, finally giving BERT a much needed do-over.
+
+ModernBERT demonstrates that encoder-only models can be improved by modern methods. They continue to offer very strong performance on some tasks, providing an extremely attractive size/performance ratio.
+
+More than anything, we’re really looking forward to seeing what creative ways to use these models the community will come up with! To encourage this, we’re opening a call for demos until January 10th, 2025: the 5 best ones will get added to this post in a showcase section and win a $100 (or local currency equivalent) Amazon gift card, as well as a 6-month HuggingFace Pro subscription! (If you need a hint to get started, here’s a demo we thought about: code similarity HF space)
```
Comment (tomaarsen) [inline modernbert.md:259]: We might want to reiterate that for any task other than mask filling, the model should be finetuned. For similarity, that could be with e.g. Sentence Transformers.

Comment (tomaarsen) [reply modernbert.md]: Resolved in 0404578b8f7df63acf338ff01099b04e4c0129dc

File: _blog.yml [hunk 17]
```diff
@@ -5215,3 +5215,14 @@
...
+- local: modernbert
+  title: "Finally, a Replacement for BERT: Introducing ModernBERT"
+  author: bclavie
+  guest: true
+  thumbnail: /blog/assets/modernbert/thumbnail.png
+  date: December 19, 2024
+  tags:
+    - nlp
+    - community
+    - research
```
Comment (tomaarsen) [inline _blog.yml:5228]:
```suggestion
    - research
    - open-source-collab
```
I think this one is fitting as well.

File: modernbert.md [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+
+## Links
+
+- [🤗ModernBERT-Base](https://huggingface.co/answerdotai/ModernBERT-base)  
+- [🤗ModernBERT-Large](https://huggingface.co/answerdotai/ModernBERT-large)  
+- [📝**arXiv**](https://arxiv.org/abs/2412.13663)  
+- [🤗ModernBERT documentation page](https://huggingface.co/docs/transformers/en/model_doc/modernbert)
```
Comment (NohTow) [inline modernbert.md:266]:
```suggestion
- [🤗ModernBERT documentation page](https://huggingface.co/docs/transformers/main/en/model_doc/modernbert)
```

**Comment by pcuenca** [comment]
> Adding code snippet

🔥
